### PR TITLE
FS3000 sensor

### PR DESCRIPTION
**Pull request by kahrendt**
# What does this implement/fix?

This pull request adds support for the FS3000 Air Velocity Sensor. It includes support for the two available models, 1005 and 1015.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes (not applicable)

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2715

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
sensor:
  - platform: fs3000
    name: "Air Velocity"
    model: 1005
    update_interval: 60s

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by SeByDocKy** [comment]
Very interesting sensor.... and thanks for the ESPhome integration.... I may use it in a big project

File: esphome/components/fs3000/sensor.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+FS3000_MODEL_OPTIONS = {
+    "1005": FS3000Model.FIVE,
+    "1015": FS3000Model.FIFTEEN,
+}
+
+FS3000Component = fs3000_ns.class_(
+    "FS3000Component", cg.PollingComponent, i2c.I2CDevice, sensor.Sensor
+)
+
+CONFIG_SCHEMA = cv.All(
```
Comment (jesserockz) [inline esphome/components/fs3000/sensor.py]: No need to wrap in `All`
```suggestion
CONFIG_SCHEMA = (
```

File: esphome/components/fs3000/sensor.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    sensor.sensor_schema(
+        FS3000Component,
+        unit_of_measurement="m/s",
+        accuracy_decimals=2,
+        device_class=DEVICE_CLASS_WIND_SPEED,
+        state_class=STATE_CLASS_MEASUREMENT,
+    )
+    .extend(
+        {
+            cv.GenerateID(): cv.declare_id(FS3000Component),
```
Comment (jesserockz) [inline esphome/components/fs3000/sensor.py]: This is not needed as the `sensor_schema()` adds it already
```suggestion
```

File: esphome/components/fs3000/sensor.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    .extend(cv.polling_component_schema("60s"))
+    .extend(i2c.i2c_device_schema(0x28))
+)
+
+
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    await cg.register_component(var, config)
+    await i2c.register_i2c_device(var, config)
+    await sensor.register_sensor(var, config)
```
Comment (jesserockz) [inline esphome/components/fs3000/sensor.py]:
```suggestion
    var = await sensor.new_sensor(config)
    await cg.register_component(var, config)
    await i2c.register_i2c_device(var, config)
```

File: esphome/components/fs3000/fs3000.cpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    std::copy(MPS_DATA_POINTS_1005, MPS_DATA_POINTS_1005 + 9, this->mps_data_points_);
+  } else if (model_ == FIFTEEN) {
+    // datasheet gives 13 points to extrapolate form for the 1015 model
+    static const uint16_t RAW_DATA_POINTS_1015[13] = {409,  1203, 1597, 1908, 2187, 2400, 2629,
+                                                      2801, 3006, 3178, 3309, 3563, 3686};
+    static const float MPS_DATA_POINTS_1015[13] = {0.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 13.0, 15.0};
+
+    std::copy(RAW_DATA_POINTS_1015, RAW_DATA_POINTS_1015 + 13, this->raw_data_points_);
+    std::copy(MPS_DATA_POINTS_1015, MPS_DATA_POINTS_1015 + 13, this->mps_data_points_);
+  }
```
Comment (jesserockz) [inline esphome/components/fs3000/fs3000.cpp]: I would prefer this initialization is done in `setup` and leave the `setter` to be just a `setter` for the `model` variable

File: esphome/components/fs3000/fs3000.h [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+ public:
+  void update() override;
+
+  void dump_config() override;
+  float get_setup_priority() const override { return setup_priority::DATA; }
+
+  void set_model(FS3000Model model);
+
+ protected:
+  FS3000Model model_{FIVE};
```
Comment (jesserockz) [inline esphome/components/fs3000/fs3000.h]: `set_model` is always called by code generation so a default is not needed in the header
```suggestion
  FS3000Model model_{};
```

File: esphome/components/fs3000/fs3000.cpp [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+  //  byte 3 - (8 bits) low byte of sensor reading
+  //  byte 4 - generic checksum data
+  //  byte 5 - generic checksum data
+
+  uint8_t data[5];
+
+  if (!this->read_bytes_raw(data, 5)) {
+    this->status_set_warning();
+    ESP_LOGW(TAG, "Error reading data from FS3000");
+    publish_state(NAN);
```
Comment (jesserockz) [inline esphome/components/fs3000/fs3000.cpp]:
```suggestion
    this->publish_state(NAN);
```

File: esphome/components/fs3000/fs3000.cpp [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    ESP_LOGW(TAG, "Checksum failure when reading from FS3000");
+    return;
+  }
+
+  // raw value information is 12 bits
+  uint16_t raw_value = (data[1] << 8) | data[2];
+  ESP_LOGV(TAG, "Got raw reading=%i", raw_value);
+
+  // convert and publish the raw value into m/s using the table of data points in the datasheet
+  publish_state(fit_raw_(raw_value));
```
Comment (jesserockz) [inline esphome/components/fs3000/fs3000.cpp]:
```suggestion
  this->publish_state(fit_raw_(raw_value));
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/fs3000/fs3000.cpp (hunk 4 above)
Comment (kahrendt) [reply esphome/components/fs3000/fs3000.cpp]: Makes sense, thanks!

**Comment by kahrendt** [comment]
Thank you for taking the time to review my pull request! This is all fairly new to me, and I appreciate your help. I made the requested changes, including the in-depth setting up code for each module into the setup() function, as well as fixed the sensor.py to remove the unnecessary components.

**Review by jesserockz — APPROVED** [review]

**Pull request merged**